pkg/worker/log/logger.go

```diff
@@ -0,0 +1,116 @@
+package log
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"syscall"
+
+	"goteleport.com/pkg/worker/conf"
+)
+
+type Logger struct {
+	config  conf.Config
+	watcher Watcher
+}
+
+// NewLogger returns a new Logger instance.
+func NewLogger(config conf.Config) Logger {
+	return Logger{
+		config:  config,
+		watcher: NewWatcher(),
+	}
+}
+
+// Path returns a absolute file path given a name.
+func (l *Logger) Path(name string) (logpath string) {
+	return filepath.Join(l.config.LogFolder(), fmt.Sprintf("%s.log", name))
+}
+
+// Create creates and return a os.File under log folder.
+// If the file can't be created an error will be returned.
+func (l *Logger) Create(name string) (f *os.File, err error) {
+	return os.Create(l.Path(name))
+}
+
+// Remove removes the named file under the log folder.
+func (l *Logger) Remove(name string) (err error) {
+	return os.Remove(l.Path(name))
+}
+
+// Tailf watch a named log file under the log folder, and
+// streams his content through a channel.
+// If the context is canceled, the tail will stop to watch the
+// log file, and the channel will be closed.
+func (l *Logger) Tailf(extctx context.Context, name string) (logchan chan string, err error) {
+	file, err := os.OpenFile(l.Path(name), os.O_RDONLY, 0644)
+	if os.IsNotExist(err) {
+		return
+	}
+	ctx, cancel := context.WithCancel(extctx)
+	// watching for modify and delete events
+	eventchan, err := l.watcher.Watch(ctx, l.Path(name), syscall.IN_MODIFY|syscall.IN_DELETE)
+	if err != nil {
+		cancel()
+		return
+	}
+	logchan = make(chan string)
+	go func() {
+		defer func() {
+			file.Close()
+			close(logchan)
+			cancel()
+		}()
+		reader := bufio.NewReader(file)
+		offset := 0
+		// stream the file from the begin
+		for {
+			// read line by line
+			bytesr, err := reader.ReadBytes('\n')
+			if err != nil {
+				break
+			}
+			select {
+			case logchan <- string(bytesr):
+			case <-extctx.Done():
+				return
+			}
+			// calculate the offset
+			offset += len(bytesr)
+		}
+		// handle file events to keep reading
+		for {
```

Comment:
I got it, for the first loop the EOF will break the iteration, but for the second one not, since we will keep the log file the goroutine will keep running. To address it I will watch the syscall.IN_CLOSE_WRITE,  syscall.IN_CLOSE and syscall.IN_CLOSE_NOWRITE events.
What do you think?